### PR TITLE
Implement background-position-x/y

### LINKED ISSUE #14458: Implement parsing/serialization for background-position-x/y properties
`background-position` is a longhand currently. We need to implement `background-position-x/y` longhands and make `background-position` a shorthand instead.

But currently [`Position`](https://doc.servo.org/style/values/specified/position/struct.Position.html) struct seems insufficient for this properties. We might need to add VerticalPosition and HorizontalPosition structs like these:
```rust
pub struct VerticalPosition {
    pub keyword: Option<VerticalKeyword>,
    pub position: Option<LengthOrPercentage>,
}
pub struct HorizontalPosition {
    pub keyword: Option<HorizontalKeyword>,
    pub position: Option<LengthOrPercentage>,
}
```
And we can change main Position struct afterwards:
```rust
pub struct Position {
    pub horizontal: Option<HorizontalPosition>,
    pub vertical: Option<VerticalPosition>,
}
```
Currently `Keyword` enum is used by both vertical and horizontal keywords. We need to add x/y-start and x/y-end as well. It might be better to split these enums but I'm not sure about it.
Well, we can reduce the structs without splitting the Keyword enums:
```rust
pub struct SubPosition {
    pub keyword: Option<Keyword>,
    pub position: Option<LengthOrPercentage>,
}
pub struct Position {
    pub horizontal: Option<SubPosition>,
    pub vertical: Option<SubPosition>,
}
```
Another option is to leave Position struct like this and make a new struct for position-x/y properties. But this is probably not a good choice.

I want to work on this, second structure looks better to me but I couldn't decide about it, any advice about this?

cc @Manishearth

### PR DESCRIPTION
**Pull request by canova**
This is a WIP PR. Just HorizontalPosition / VerticalPosition implementations are complete. I would like to get early feedbacks about this architecture. Here's some architectural topics to consider:

- I created `HorizontalPosition` and `VerticalPosition` structs for this and used them in `Position` as well. We have decided to split `Keyword` enum, but we need them as unified for `PositionComponent` enum. So I didn't split but I can split it if we prefer to change PositionComponent as well.
- If we prefer Keyword enum like this, we can create a SubPosition(or something like this) instead of HorizontalPosition/VerticalPosition enums since only difference is 2 lines in `parse` functions. We can create a `parse_horizontal` and `parse_vertical` instead and a lot of code duplication can be cleared.
- I couldn't find a good way to use HorizontalPosition/VerticalPosition's parse functions in `Position`'s parse function. It is a bit more complicated. I'm open to suggestions :)
- I don't know much about logical keywords so do I need to do something different? I placed some comments where logical keywords are processing.

Any advice about these?

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14458 (github issue number if applicable).

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14509)

### CODE REVIEW

**Comment by canaltinova** [comment]
And I made unusable `x/y-start`, `x/y-end` keywords for `Position`. They are not valid in current spec yet.

File: components/style/values/computed/position.rs [hunk 1]
```diff
@@ -26,3 +26,19 @@ impl ToCss for Position {
...
+    }
+}
+
+pub struct VerticalPosition(pub LengthOrPercentage);
+
+impl ToCss for VerticalPosition {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        self.0.to_css(dest)
+    }
+}
```
Comment (Manishearth) [inline components/style/values/computed/position.rs]: nit: newline at EOF

**Review by Manishearth — COMMENTED** [review]
> If we prefer Keyword enum like this, we can create a SubPosition(or something like this) instead of HorizontalPosition/VerticalPosition enums

Perhaps just have `Position<Horizontal>` and `Position<Vertical>` using empty structs `Horizontal` and `Vertical` (with different parse impls on just the H/V structs, and have the Position type call out to these.

> I don't know much about logical keywords so do I need to do something different? 

They basically parse based on the writing mode of the context when computing. Ignore for now, but file a bug. I'll give you pointers for fixing it later.

File: components/style/values/computed/position.rs (hunk 1 above)
Comment (Wafflespeanut) [inline components/style/values/computed/position.rs]: There's a newline, I guess? Github doesn't show it. It only shows when there "isn't" one :smile:

**Comment by canaltinova** [comment]
@Manishearth Hm, that looks better. I'll do that, thanks! :)

**Comment by canaltinova** [comment]
@Manishearth Hm, that looks better. I'll do that, thanks! :)
(Oops, closed by mistake)

**Comment by canaltinova** [comment]
I think it's ready to review right now :)

File: components/style/properties/gecko.mako.rs [hunk 2]
```diff
@@ -1635,6 +1639,57 @@ fn static_assert() {
...
+        self.gecko.mImage.mLayers.mFirstElement.mPosition =
+            other.gecko.mImage.mLayers.mFirstElement.mPosition;
+        unsafe {
+            Gecko_EnsureImageLayersLength(&mut self.gecko.mImage,
+                                          other.gecko.mImage.mLayers.len(),
+                                          LayerType::Background);
+        }
+        for (layer, other) in self.gecko.mImage.mLayers.iter_mut()
+                                  .zip(other.gecko.mImage.mLayers.iter())
+                                  .take(other.gecko.mImage.mPosition${orientation[0].upper()}Count as usize) {
```
Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: take shouldn't be necessary here afaict.

File: components/style/properties/longhand/background.mako.rs [hunk 3]
```diff
@@ -86,43 +86,73 @@ ${helpers.predefined_type("background-color", "CSSColor",
...
+            use values::specified::position::Keyword;
+            HorizontalPosition {
+                keyword: Some(Keyword::Left),
+                position: None,
             }
         }
+
+        pub fn parse(context: &ParserContext, input: &mut Parser)
+                     -> Result<SpecifiedValue, ()> {
+            Ok(try!(HorizontalPosition::parse(context, input)))
```
Comment (Manishearth) [inline components/style/properties/longhand/background.mako.rs]: `Ok(try!(foo))` at the end of a function is just `foo`

**Review by Manishearth — APPROVED** [review]
Minor nits, lgtm

@bors-servo delegate+

**Comment by canaltinova** [comment]
@bors-servo r=Manishearth
thanks!

**Comment by canaltinova** [comment]
Oops

**Comment by KiChjang** [comment]
@bors-servo retry
* git gud

**Comment by KiChjang** [comment]
@bors-servo retry
* git gud

**Comment by KiChjang** [comment]
```
Ran 9977 tests finished in 1439.0 seconds.
  • 9973 ran as expected. 147 tests skipped.
  • 1 tests timed out unexpectedly
  • 3 tests had unexpected subtest results

Tests with unexpected results:
  ▶ TIMEOUT [expected PASS] /css-text-decor-3_dev/html/text-emphasis-style-property-010Cn.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ assertion failed: glyphs[0] != 0 (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /Users/servo/buildbot/slave/mac-rel-css/build/components/gfx/platform/macos/font.rs:267)
  │ stack backtrace:
  │    0:        0x106d190de - backtrace::backtrace::trace::h90be2828c7ec6b59
  │    1:        0x106d193cc - backtrace::capture::Backtrace::new::h8135b43a33e0c535
  │    2:        0x106b904d4 - servo::main::_{{closure}}::h43e3ad31baaf2128
  │    3:        0x10867eb43 - std::panicking::rust_panic_with_hook::ha5aed1dfc0e220e3
  │    4:        0x1080e27d4 - std::panicking::begin_panic::h5f0a1fd8f4de57ec
  │    5:        0x108101cd7 - _<gfx..platform..macos..font..FontHandle as gfx..font..FontHandleMethods>::glyph_index::h9e95c651a13529f5
  │    6:        0x1080bb7ea - layout::text::TextRunScanner::scan_for_runs::hb2774506ed011cfe
  │    7:        0x106ddfdc4 - _<layout..construct..FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block_starting_with_fragments::h6470c950bac09094
  │    8:        0x106ddce3e - _<layout..construct..FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block_like::h08f51919a1c6efe0
  │    9:        0x106ddc3a4 - _<layout..construct..FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block::h7a10a1d01a5a7817
  │   10:        0x106d88d6a - _<layout..construct..FlowConstructor<'a, ConcreteThreadSafeLayoutNode> as layout..traversal..PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>>::process::h296e5ca81b10c1c0
  │   11:        0x106d7ffd2 - _<layout..traversal..RecalcStyleAndConstructFlows<'lc> as style..traversal..DomTraversalContext<N>>::process_postorder::hff626f47eb44bc15
  │   12:        0x106dbbeb8 - style::sequential::traverse_dom::doit::ha0ce22425f5c2e13
  │   13:        0x106dbbe61 - style::sequential::traverse_dom::doit::ha0ce22425f5c2e13
  │   14:        0x106dbbe61 - style::sequential::traverse_dom::doit::ha0ce22425f5c2e13
  │   15:        0x106dfa51e - layout_thread::LayoutThread::handle_reflow::h3b435401362fb54c
  │   16:        0x106df400b - layout_thread::LayoutThread::handle_request_helper::h39309b0025b7dedc
  │   17:        0x106defd3d - layout_thread::LayoutThread::handle_request::ha37a7cbfbce96723
  │   18:        0x106dea06a - _<layout_thread..LayoutThread as layout_traits..LayoutThreadFactory>::create::_{{closure}}::h2855f63d876c3013
  │   19:        0x106d9d097 - std::panicking::try::do_call::h3367fddee8aa42e1
  │   20:        0x10867fa1a - __rust_maybe_catch_panic
  │   21:        0x106db53f4 - _<F as alloc..boxed..FnBox<A>>::call_box::hcd88456dc30e3e16
  │   22:        0x10867dac4 - std::sys::imp::thread::Thread::new::thread_start::h8084b1107992ae5b
  │   23:     0x7fff97f2c059 - _pthread_body
  │   24:     0x7fff97f2bfd6 - _pthread_start
  │ ERROR:servo: assertion failed: glyphs[0] != 0
  │ called `Result::unwrap()` on an `Err` value: "SendError(..)" (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at ../src/libcore/result.rs:837)
  │ stack backtrace:
  │    0:        0x106d190de - backtrace::backtrace::trace::h90be2828c7ec6b59
  │    1:        0x106d193cc - backtrace::capture::Backtrace::new::h8135b43a33e0c535
  │    2:        0x106b904d4 - servo::main::_{{closure}}::h43e3ad31baaf2128
  │    3:        0x10867eb43 - std::panicking::rust_panic_with_hook::ha5aed1dfc0e220e3
  │    4:        0x10867ea14 - std::panicking::begin_panic::hc066339e2fdc17d1
  │    5:        0x10867e932 - std::panicking::begin_panic_fmt::h5912b2d2df332044
  │    6:        0x10867e897 - rust_begin_unwind
  │    7:        0x1086ac750 - core::panicking::panic_fmt::h561c5ee168a3d2cb
  │    8:        0x1070ea3b0 - core::result::unwrap_failed::h88e9aa51a446bca0
  │    9:        0x1073adffe - script::dom::window::Window::force_reflow::h55666464c8ae5bec
  │   10:        0x10742222e - script::script_thread::ScriptThread::handle_event::h9727fe9bb5ce1589
  │   11:        0x1073f27b2 - script::script_thread::ScriptThread::handle_msgs::hd4327a08dc341491
  │   12:        0x10704d1b7 - std::panicking::try::do_call::hc514376262ffc812
  │   13:        0x10867fa1a - __rust_maybe_catch_panic
  │   14:        0x107113c96 - _<F as alloc..boxed..FnBox<A>>::call_box::h786c706be160262d
  │   15:        0x10867dac4 - std::sys::imp::thread::Thread::new::thread_start::h8084b1107992ae5b
  │   16:     0x7fff97f2c059 - _pthread_body
  │   17:     0x7fff97f2bfd6 - _pthread_start
  └ ERROR:servo: called `Result::unwrap()` on an `Err` value: "SendError(..)"

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-331.htm:
  │ FAIL [expected PASS] background_initial_position
  │   → assert_equals: background initial value for background-position expected "0% 0%" but got ""
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-331.htm:10:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-331.htm:9:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-333.htm:
  │ FAIL [expected PASS] background_specified_color_position
  │   → assert_equals: background initial value for background-position expected "0% 0%" but got ""
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-333.htm:10:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-333.htm:9:9

  ▶ Unexpected subtest result in /css-transitions-1_dev/html/changing-while-t
```

**Pull request merged**